### PR TITLE
feature - get device token

### PR DESCRIPTION
**Pull request by at-duytran**

### CODE REVIEW

**Comment by at-daiho** [comment]
:+1: There are no linter violations.

File: MyApp/View/AppDelegate.swift [hunk 1]
```diff
@@ -36,6 +36,10 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
         configDatabase()
         configIQKeyboard()
         configRealm()
+        let notificationTypes: UIUserNotificationType = [.alert, .badge, .sound]
+        let pushNotificationSettings = UIUserNotificationSettings(types: notificationTypes, categories: nil)
+        application.registerUserNotificationSettings(pushNotificationSettings)
+        application.registerForRemoteNotifications()
```
Comment (at-truongnguyen) [inline MyApp/View/AppDelegate.swift]: đưa cái ni ra thành hàm riêng

File: MyApp/View/AppDelegate.swift [hunk 2]
```diff
@@ -49,6 +53,75 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
         checkAuth()
         return true
     }
+}
+
+// MARK: Extension AppDelegate, function
```
Comment (at-truongnguyen) [inline MyApp/View/AppDelegate.swift]: Update meaning of this mark.

File: MyApp/View/AppDelegate.swift [hunk 3]
```diff
@@ -49,6 +53,75 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
...
+
+                    if let rootSideMenu = rootSideMenu as? UINavigationController, let topVC = rootSideMenu.topViewController {
+                        let vc = DetailOrderViewController()
+                        vc.viewModel = DetailOrderViewModel(orderId: id)
+                        topVC.navigationController?.pushViewController(vc, animated: true)
+                    }
+                }
+            }
+        } else {
+            logOut()
```
Comment (at-truongnguyen) [inline MyApp/View/AppDelegate.swift]: why we have to logout if cant not get device token?

File: MyApp/View/Controllers/Login/LoginViewController.swift [hunk 4]
```diff
@@ -62,6 +62,19 @@ final class LoginViewController: ViewController {
...
                 case .success:
+                    let userDefaults = UserDefaults.standard
+                    if let deviceToken = userDefaults.string(forKey: UserDefaults.Key.deviceToken) {
+                        this.viewModel.sendingDeviceToken(deviceToken: deviceToken) { (result) in
+                            switch result {
+                            case .success:
+                                print("-> this.viewModel.sendingDeviceToken: \(deviceToken)")
+                                break
+                            case .failure:
+                                break
```
Comment (at-truongnguyen) [inline MyApp/View/Controllers/Login/LoginViewController.swift]: remove unused code

File: MyApp/Model/API/API.DeviceToken.swift [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        let path = Api.Path.DeviceToken()
+        return api.request(method: .post, urlString: path, parameters: params.toJSON()) { (result) in
+            switch result {
+            case .success:
+                DispatchQueue.main.async {
+                    completion(.success(result))
+                }
+            case .failure(let error):
+                completion(.failure(error))
+            }
```
Comment (at-truongnguyen) [inline MyApp/Model/API/API.DeviceToken.swift]: Do you really need to check the result?

File: MyApp/View/AppDelegate.swift [hunk 6]
```diff
@@ -36,6 +36,7 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
         configDatabase()
         configIQKeyboard()
         configRealm()
+        registerForPushNotification()
```
Comment (at-truongnguyen) [inline MyApp/View/AppDelegate.swift]: `registerRemoteNotification` please

File: MyApp/View/AppDelegate.swift [hunk 7]
```diff
@@ -49,6 +50,80 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
...
+        let token = UserDefaults.standard.string(forKey: UserDefaults.Key.accessToken)
+
+        if let _ = token {
+            guard let _ = userInfo["type"] as? String, let id = userInfo["id"] as? Int else {
+                return
+            }
+
+            switch application.applicationState {
+            case .active:
+                NotificationCenter.default.post(name: NSNotification.Name(App.Strings.unreadNotification), object: nil)
```
Comment (at-truongnguyen) [inline MyApp/View/AppDelegate.swift]: change notification name

File: MyApp/View/Controllers/Login/LoginViewController.swift [hunk 8]
```diff
@@ -62,6 +62,18 @@ final class LoginViewController: ViewController {
...
                 case .success:
+                    let userDefaults = UserDefaults.standard
+                    if let deviceToken = userDefaults.string(forKey: UserDefaults.Key.deviceToken) {
+                        this.viewModel.sendingDeviceToken(deviceToken: deviceToken) { (result) in
+                            switch result {
+                            case .success:
+                                break
+                            case .failure:
+                                break
+                            }
```
Comment (at-truongnguyen) [inline MyApp/View/Controllers/Login/LoginViewController.swift]: code thừa mà cũng code 😞

File: MyApp/View/Controllers/Menu/Cell/SideMenuTableCell.swift [hunk 9]
```diff
@@ -38,8 +38,21 @@ final class SideMenuTableCell: TableCell {
...
+            }
+        }
+
+        if let notificationBadgeNumber = viewModel.notificationBadgeNumber {
+            if notificationBadgeNumber > 0 {
+                badgeLabel.text = "\(notificationBadgeNumber)"
+                badgeLabel.backgroundColor = #colorLiteral(red: 0.9254902005, green: 0.2352941185, blue: 0.1019607857, alpha: 1)
+            } else {
+                badgeLabel.text = ""
+            }
```
Comment (at-truongnguyen) [inline MyApp/View/Controllers/Menu/Cell/SideMenuTableCell.swift]: xử lý `if` trong view model chớ hè?

File: MyApp/View/Controllers/Menu/Cell/SideMenuTableCell.swift [hunk 10]
```diff
@@ -38,8 +38,21 @@ final class SideMenuTableCell: TableCell {
...
 
         if let badgeNumber = viewModel.badgeNumber {
-            badgeLabel.text = "\(badgeNumber)"
-            badgeLabel.backgroundColor = #colorLiteral(red: 0.3943337798, green: 0.7393168807, blue: 0.1809048057, alpha: 1)
+            if badgeNumber > 0 {
+                badgeLabel.text = "\(badgeNumber)"
+                badgeLabel.backgroundColor = #colorLiteral(red: 0.3943337798, green: 0.7393168807, blue: 0.1809048057, alpha: 1)
+            } else {
+                badgeLabel.text = ""
+            }
```
Comment (at-truongnguyen) [inline MyApp/View/Controllers/Menu/Cell/SideMenuTableCell.swift]: xử lý trong view model chớ hè? trên view chỉ có lấy ra hiển thị thôi chớ?

File: MyApp/View/Controllers/Menu/MenuViewController.swift [hunk 11]
```diff
@@ -11,10 +11,10 @@ import UIKit
 final class MenuViewController: ViewController {
 
     // MARK: - Properties
-    @IBOutlet private weak var userAvatarImage: UIImageView!
-    @IBOutlet private weak var userNameLabel: UILabel!
-    @IBOutlet private weak var userEditButton: UIButton!
-    @IBOutlet private weak var tableView: UITableView!
+    @IBOutlet fileprivate weak var userAvatarImage: UIImageView!
```
Comment (at-truongnguyen) [inline MyApp/View/Controllers/Menu/MenuViewController.swift]: check your variable name

File: MyApp/ViewModel/Login/LoginViewModel.swift [hunk 12]
```diff
@@ -77,11 +77,30 @@ final class LoginViewModel: MVVM.ViewModel {
...
+    typealias SendingDeviceTokenCompletion = (SendingDeviceTokenResult) -> Void
+    func sendingDeviceToken(deviceToken: String, _ completion: @escaping SendingDeviceTokenCompletion) {
+        let deviceTokenParam = Api.DeviceToken.QueryParams(deviceToken: deviceToken)
+        Api.DeviceToken.queryDeviceToken(params: deviceTokenParam) { (result) in
+            switch result {
+            case .success(_):
+                completion(.success)
+            case .failure:
+                completion(.failure)
+            }
```
Comment (at-truongnguyen) [inline MyApp/ViewModel/Login/LoginViewModel.swift]: thừa

File: MyApp/View/Base/ViewController.swift [hunk 13]
```diff
@@ -20,7 +20,7 @@ class ViewController: UIViewController, MVVM.View {
 
         setupNavigationBar()
         NotificationCenter.default.addObserver(self, selector: #selector(updateNotiNavigationBar),
-                                               name: NSNotification.Name(rawValue: App.Strings.unreadNotification),
+                                               name: .didReceiveRemoveNotification,
```
Comment (at-truongnguyen) [inline MyApp/View/Base/ViewController.swift]: remote

File: MyApp/View/Controllers/Menu/MenuViewController.swift [hunk 14]
```diff
@@ -11,9 +11,9 @@ import UIKit
 final class MenuViewController: ViewController {
 
     // MARK: - Properties
-    @IBOutlet fileprivate weak var userAvatarImage: UIImageView!
+    @IBOutlet fileprivate weak var userProfileImage: UIImageView!
```
Comment (at-truongnguyen) [inline MyApp/View/Controllers/Menu/MenuViewController.swift]: naming

**Pull request closed without merging**